feat: CustomException 을 계층화한다

GuruneLee opened the pull request:
post 도메인에서 던지는 예외를 모두 클래스화 하였습니다.
부모 class 로 'PostException' 을 두고 이를 상속하여 구현하였습니다.
각 도메인 별 던지는 예외는 [여기](https://cut-table-15f.notion.site/untitled-88b7b334eafe42ffb93587bb2816d3d0)에 정리하였습니다.

메시지와 http code를 미리 정의해 두는 것에 의의가 있는 것 같습니다.

문제점은 다음과 같습니다.
1. post 도메인에서 뱉는다고 꼭 PostException 의 자식으로 하면 어색한 부분이 있습니다. (ex. NonExistentUser)
2. 각 도메인이 뱉는 '상황'을 기준으로 하면, 재사용성이 떨어집니다 (ex. creatAnswer 메서드를 호출할 시, post가 없어 실패하면 CanNotCreateAnswer 예외를 뱉어야 합니까? 아니면 NonExistentPost 예외를 뱉어야 합니깝숑?)

첫 커밋은 코드 자체보단 아이디어를 나누고 싶습니다.

Close #75

GuruneLee committed:
feat: custom exception 중 post 도메인에서 던지는 예외를 클래스화 하였다.

wannte commented on src/main/java/com/example/gistcompetitioncnserver/exception/ControllerAdvice.java:
```diff
@@ -13,6 +14,11 @@
                 .body(new ErrorMessage(400, ex.getMessage()));
     }
 
+    @ExceptionHandler(PostException.class)
```
도메인 별로 별도의 ExceptionHandler를 작성해주어야 할거 같은데요!

해당 도메인 Exception들의 부모클래스를 두고, 하나로 처리하는 방법은 어떨까요?

도메인이 추가될 수록, ExceptionHandler가 추가되는 부분은 불편할 수 있겠다는 생각입니다!

GuruneLee replied:
네 저도 루트 예외를 하나 두는것을 염두해두고 있었습니다

wannte commented on src/main/java/com/example/gistcompetitioncnserver/exception/post/PostException.java:
```diff
@@ -0,0 +1,17 @@
+
+import org.springframework.http.HttpStatus;
+
+public class PostException extends RuntimeException {
```
도메인 별로 Exception을 분리하게 되었을 때 어떤 장점을 가질 수 있을까요?
또 User가 존재하는지에 대한 테스트를 거의 대부분의 서비스에서 진행하게 될거 같은데, 각 도메인 별로 별도로 정의해주는 부분은 비효율적일수도 있겠다는 생각인데요! 이부분의 생각도 궁금합니다!

이러한 공통적인 부분은 common 정도의 패키지에 묶어서 관리할 수도 있겠네요!

wannte commented on src/main/java/com/example/gistcompetitioncnserver/exception/post/DuplicatedAgreementException.java:
```diff
@@ -0,0 +1,11 @@
+public class DuplicatedAgreementException extends PostException {
+    private static final HttpStatus HTTP_STATUS = HttpStatus.METHOD_NOT_ALLOWED;
+    private static final String MESSAGE = "동의는 한 번만 할 수 있습니다";
+    public DuplicatedAgreementException() {
```
공백 또한 컨벤션이니 한줄 띄어주심이!

wannte reviewed (commented):
빠르게 진행해주셨네요!
간단하게 몇가지 질문 남기고 갑니다! 요 부분에 대해서만, 답변 진행해주시고 이어서 진행해도 좋을 것 같아요!

ChoiEungi reviewed (commented):
domain 별로 문서화까지 너무 고생하셨습니다! notion의 내용을 wiki로 옮겨도 좋을 것 같습니다!

GuruneLee committed:
revert: 테스트용 post exception 들을 삭제한다

GuruneLee committed:
feat: ApplicationException 을 정의한다

GuruneLee committed:
feat: exception/post 를 정의한다

GuruneLee committed:
feat: exception/user 를 정의한다

GuruneLee committed:
feat: exception/comment 를 정의한다

GuruneLee committed:
feat: post 도메인에 계층화된 exception 을 적용한다

GuruneLee committed:
refactor: UnAuthenticatedUserException 의 이름을 NotConfirmedEmailException 으로 바꾼다

GuruneLee committed:
feat: answer 도메인에 계층화된 exception 을 적용한다

GuruneLee committed:
feat: comment 도메인에 계층화된 exception 을 적용한다

GuruneLee committed:
feat: user 도메인에 계층화된 exception 을 적용한다

GuruneLee committed:
refactor: validation 에 계층화된 exception 도입

GuruneLee committed:
refactor: ControllerAdvice 에 ApplicationException handler 구현

GuruneLee committed:
refactor: CustomException 수정

CustomException 이 CheckedException 과 메시지, httpStatus 를 담을 수 있도록 수정. 기본 HttpStatus 는 400 (Bad Request)

GuruneLee committed:
feat: RegistrationListener 에 계층화된 exception 도입

GuruneLee committed:
chore: import 추가

GuruneLee committed:
feat: AnswerServiceTest 에 계층화된 exception 추가

GuruneLee committed:
feat: 변경된 CustomException 적용

GuruneLee committed:
refactor: VerificationServiceTest 에 놓친 exception 반영

GuruneLee committed:
refactor: post 도메인에 적용된 exception 세분화

GuruneLee committed:
Merge branch 'develop' into feature/custom-exception

GuruneLee committed:
fix: after resolving conflicts...

koseyeon commented on src/main/java/com/example/gistcompetitioncnserver/exception/CustomException.java:
```diff
@@ -1,7 +1,17 @@
+    public CustomException(String message, Throwable cause) {
+        this(message, cause, HttpStatus.BAD_REQUEST);
+    }
+    public CustomException(String message, Throwable cause, HttpStatus httpStatus) {
```
다른 exception들은 다 분류를 하셨는데 (postException 등) ,
이메일익셉션, 파일익셉션들은 분류를 따로 안하고 customException으로 남기신 이유가 궁금합니다!

wannte replied:
CustomException과 ApplicationException의 기준이 조금 더 명확했으면 좋겠다는 생각입니다! (클래스 명에서 그 역할이 명시적으로 들어났으면 합니다)

CustomExcepion -> 우리가 커스텀하게 정의한 Exception
ApplicationException -> 어플리케이션을 동작시키는 데 발생하는 Exception

요 정도로 그 이름만을 봤을 떄 유추할 수 있을 것 같아요!

이러한 관점에서 보면 이메일익셉션과 파일익센션도 모두 CustomException, ApplicationException에 포함되도 어색하지 않을 것 같다는 생각입니다.

CustomException과 ApplicationException을 동시에 두었을 때 어떠한 장점을 가질 수 있는지에 대한 생각을 들어보고 싶습니다!

wannte commented on src/test/java/com/example/gistcompetitioncnserver/answer/AnswerServiceTest.java:
```diff
@@ -78,7 +80,7 @@ void retrieveAnswer() {
         assertThatThrownBy(
                 () -> answerService.retrieveAnswerByPostId(savedPost.getId())
-        ).isInstanceOf(CustomException.class);
+        ).isInstanceOf(UnAnsweredPostException.class);
```
테스트 메소드 명을 수정했으면 좋겠습니다!
제안: retrieveAnswerFromNotAnsweredPost()
Answer되지 않은 경우인데, 이 부분이 들어나지 않아 헷갈렸네요!(근데 아마 이거 내가 바꾼거 같기도 하고..)

밑에 메소드 명 중  `Exitent`의 명사형 보다는 `Existing`의 형용사를 쓰는 것 또한 제한 해봅니다!

wannte commented on src/main/java/com/example/gistcompetitioncnserver/answer/AnswerController.java:
```diff
@@ -24,10 +25,11 @@
         }
         if (!sessionUser.hasManagerAuthority()) {
-            throw new CustomException("답변 권한이 없습니다.");
+            throw new UnAuthorizedUserException();
+
```
공백 또한 컨벤션!

wannte commented on src/main/java/com/example/gistcompetitioncnserver/answer/AnswerController.java:
```diff
@@ -48,10 +50,11 @@
         }
         if (!sessionUser.hasManagerAuthority()) {
-            throw new CustomException("답변 수정 권한이 없습니다.");
+            throw new UnAuthorizedUserException();
+
```
공백 또한 컨벤션 2

wannte commented on src/main/java/com/example/gistcompetitioncnserver/comment/CommentService.java:
```diff
@@ -60,19 +64,20 @@ public void deleteComment(Long commentId) {
         Comment comment = findCommentById(commentId);
         if (!comment.getUserId().equals(eraserId)) {
-            throw new CustomException("댓글 삭제 권한이 없습니다");
+            throw new UnAuthorizedUserException();
+
```
공백33

wannte commented on src/main/java/com/example/gistcompetitioncnserver/exception/comment/NoSuchCommentException.java:
```diff
@@ -0,0 +1,11 @@
+
+public class NoSuchCommentException extends CommentException {
+    private static final String MESSAGE = "존재하지 않는 댓글입니다.";
+    private static final HttpStatus HTTP_STATUS = null;
```
이 부분은 내일 하나씩 보면서 결정해도 좋겠네요!

wannte commented on src/main/java/com/example/gistcompetitioncnserver/user/UserRole.java:
```diff
@@ -13,7 +13,7 @@ public static UserRole ignoringCaseValueOf(String name) {
             return valueOf(name.toUpperCase(Locale.ROOT));
         } catch (IllegalArgumentException e) {
-            throw new CustomException("존재하지 않는 유저 권한입니다.");
+            throw new CustomException("존재하지 않는 유저 권한입니다.", e);
```
이 부분을 CustomException으로 정의한 이유가 궁금합니다!

wannte commented on src/main/java/com/example/gistcompetitioncnserver/verification/VerificationService.java:
```diff
@@ -30,15 +32,15 @@ public String createToken(User user) {
     public void confirm(String token) {
         VerificationToken verificationToken = verificationTokenRepository.findByToken(token)
-                .orElseThrow(() -> new CustomException("존재하지 않는 토큰입니다."));
+                .orElseThrow(() -> new CustomException("존재하지 않는 토큰입니다.", null));
```
이 부분도 마찬가지구요!

wannte reviewed (changes requested):
Exception을 상세하게 나누고 처리하니 가독성이 훨씬 좋아진 것 같네요!

테스트에서 또한, CustomExcption이 발생한다의 검증보다도 구체적인 에러 검증을 진행할 수 있었던 것 같습니다 :+1:

한 가지, CustomException과 ApplicationException의 동시 사용에 관한 피드백을 남겼으니 확인 부탁드려요!

GuruneLee committed:
Merge branch 'develop' into feature/custom-exception

# Conflicts:
#	src/main/java/com/example/gistcompetitioncnserver/answer/AnswerController.java
#	src/main/java/com/example/gistcompetitioncnserver/comment/CommentController.java
#	src/main/java/com/example/gistcompetitioncnserver/post/PostController.java
#	src/main/java/com/example/gistcompetitioncnserver/user/UserController.java
#	src/main/java/com/example/gistcompetitioncnserver/user/UserService.java
#	src/main/java/com/example/gistcompetitioncnserver/verification/RegistrationListener.java
#	src/main/java/com/example/gistcompetitioncnserver/verification/VerificationService.java
#	src/test/java/com/example/gistcompetitioncnserver/user/UserServiceTest.java
#	src/test/java/com/example/gistcompetitioncnserver/verification/VerificationServiceTest.java

GuruneLee committed:
refactor: remove ErrorCase.java, ErrorMessage.java

GuruneLee committed:
refactor: rename CustomException to WrappedException & let WrappedException inherit ApplicationException

GuruneLee committed:
feat: add exception group of verification domain

GuruneLee committed:
feat: exception 에 http status code 추가

GuruneLee committed:
feat: NoSessionException 추가, 각 Controller에 적용

wannte commented on src/main/java/com/example/gistcompetitioncnserver/exception/user/NoSessionException.java:
```diff
@@ -0,0 +1,12 @@
+
+import org.springframework.http.HttpStatus;
+
+public class NoSessionException extends UserException {
+    private static final String MESSAGE = "세션유저가 존재하지 않습니다.";
```
'세션'을 사용하는 부분을 사용자에게 노출할 필요성이 있을까? 에 대한 고민이 남네요

UnAuthenticatedExeption, "인증되지 않은 사용자입니다" 정도의 클래스명과 메시지를 제안해봅니다.

GuruneLee replied:
UnAuthenticatedExeption, "인증되지 않은 사용자입니다" 로 바꾸는데 찬성합니다.

이를 어떻게 전달할지는 다음 이슈에서 생각해보면 좋을 것 같습니다.

wannte commented on src/test/java/com/example/gistcompetitioncnserver/answer/AnswerServiceTest.java:
```diff
@@ -47,22 +49,22 @@ void setup() {
         Long savedAnswer = answerService.createAnswer(savedPost.getId(), ANSWER_REQUEST, manager.getId());
 
-        Answer answer = answerRepository.findById(savedAnswer).orElseThrow(() -> new CustomException("존재하지 않는 answer입니다.;"));
+        Answer answer = answerRepository.findById(savedAnswer).orElseThrow(() -> new WrappedException("존재하지 않는 answer입니다.", null));
```
이 부분에서 WrappedException을 사용하는 이유가 궁금합니다.

GuruneLee replied:
answer 는 post 종속적이라 answer exception 을 따로 분리하지 않았습니다. (그럴 필요도 없었습니다)
그래서 테스트 메서드에서만 사용할 목적으로 exception 을 만드는 것이 낭비라 판단했습니다.

wannte reviewed (approved):
여러가지로 세세하게 고민할 부분이 많았던 기능이였을 거 같은데, 깔끔하게 진행해주셨네요!

몇 가지 코멘트 남겼고, 위의 코멘트 들도 확인 부탁드립니다! 리뷰는 제안일 뿐, 어떤 방법을 선택할 지는 알아서 선택해주시고 머지 진행해주시면 감사하겠습니다. 가능하면 내일 오전 중으로 진행해주셔야, 다음 feature 진행할 수 있을 것 같습니다.

어떤 결정을 내렸는지에 대한 이유나 근거를 리뷰로 남겨주시면 감사하겠습니다 👍
